File: code/17.letter-combinations-of-a-phone-number.java
/*
 * @lc app=leetcode id=17 lang=java
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    List<String> res = new ArrayList<>();
    StringBuffer sb = new StringBuffer();
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.length() == 0) {
            return res;
        }
        String[] numString = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        backtracking(0, digits, numString);
        return res;
    }

    public void backtracking(int index, String digits, String[] numString) {
        if (index == digits.length()) {
            res.add(sb.toString());
            return;
        }

        String str = numString[digits.charAt(index) - '0'];
        for (int i = 0; i < str.length(); i++) {
            sb.append(str.charAt(i));
            backtracking(index + 1, digits, numString);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}
// @lc code=end

File: code/17.letter-combinations-of-a-phone-number.go
/*
 * @lc app=leetcode id=17 lang=golang
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
func letterCombinations(digits string) []string {
	res, path := make([]string, 0), make([]byte, 0)
	m := []string{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}

	if digits == "" {
		return res
	}

	var backtracking func(index int)
	backtracking = func(index int) {
		if len(path) == len(digits) {
			tmp := string(path)
			res = append(res, tmp)
			return
		}
		digit := int(digits[index] - '0')
		str := m[digit-2]
		for i := 0; i < len(str); i++ {
			path = append(path, str[i])
			backtracking(index + 1)
			path = path[:len(path)-1]
		}
	}
	backtracking(0)
	return res
}

// @lc code=end

